cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java

```diff
@@ -203,6 +204,32 @@ public long getProgramRunCount(ProgramId programId) throws Exception {
       .collect(Collectors.toList());
   }
 
+  /**
+   * Get the latest run within the specified start and end times for the specified program.
+   *
+   * @param programId the program to get runs for
+   * @param programRunStatus status of runs to return
+   * @param start earliest start time of runs to return
+   * @param end latest start time of runs to return
+   * @param limit the maximum number of runs to return
+   * @return the latest runs for the program sorted by start time, with the newest run as the first run
+   * @throws NotFoundException if the application to which this program belongs was not found or the program is not
+   *                           found in the app
+   * @throws UnauthorizedException if the principal does not have access to the program
+   * @throws Exception if there was some other exception performing authorization checks
+   */
+  public List<RunRecord> getRuns(ProgramId programId, ProgramRunStatus programRunStatus,
+                                 long start, long end, int limit) throws Exception {
+    AuthorizationUtil.ensureAccess(programId, authorizationEnforcer, authenticationContext.getPrincipal());
+    ProgramSpecification programSpec = getProgramSpecificationWithoutAuthz(programId);
```

Comment:
I think we are having this check existence and then perform the get in many places, I think it is better we can combine these in one transaction.